### PR TITLE
Add back Socket.Disconnect, Socket.Begin/EndDisconnect, Socket.DisconnectAsync

### PR DESCRIPTION
**Pull request by ericeil**
These can only be supported on Windows, as Unix has no equivalent concept.  Ported the product code and tests from the CLR.  Part of #11788.

@stephentoub @CIPop @davidsh

### CODE REVIEW

**Comment by ericeil** [comment]
@dotnet-bot Test Outerloop Windows_NT Debug please
@dotnet-bot Test Outerloop Ubuntu14.04 Debug please
@dotnet-bot Test Outerloop OSX Debug please

File: src/System.Net.Sockets/tests/FunctionalTests/DisconnectTest.cs [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+            Assert.True(Capability.IPv4Support() || Capability.IPv6Support());
+        }
+        public void OnCompleted(object sender, SocketAsyncEventArgs args)
+        {
+            EventWaitHandle handle = (EventWaitHandle)args.UserToken;
+            handle.Set();
+        }
+
+        [Fact]
+        [OuterLoop]
```
Comment (davidsh) [inline src/System.Net.Sockets/tests/FunctionalTests/DisconnectTest.cs]: You should either put a comment on this [OuterLoop] to say why it is outer loop.  Or, better, use the new attribute for [OuterLoop] that specifies a reason that it is outer loop.

See #11345

File: src/System.Net.Sockets/tests/FunctionalTests/DisconnectTest.cs [hunk 2]
```diff
@@ -0,0 +1,165 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+// TODO: make sure everything's OuterLoop and Windows-only!
```
Comment (ericeil) [inline src/System.Net.Sockets/tests/FunctionalTests/DisconnectTest.cs]: I'll remove this comment. 😄

File: src/System.Net.Sockets/tests/FunctionalTests/DisconnectTest.cs (hunk 1 above)
Comment (ericeil) [inline src/System.Net.Sockets/tests/FunctionalTests/DisconnectTest.cs]: Thanks for reminding me.  Fixed.

File: src/System.Net.Sockets/tests/FunctionalTests/DisconnectTest.cs (hunk 2 above)
Comment (ericeil) [inline src/System.Net.Sockets/tests/FunctionalTests/DisconnectTest.cs]: Done.

**Comment by ericeil** [comment]
@dotnet-bot Test Outerloop Windows_NT Debug please
@dotnet-bot Test Outerloop Ubuntu14.04 Debug please
@dotnet-bot Test Outerloop OSX Debug please

**Comment by ericeil** [comment]
@dotnet-bot test Innerloop CentOS7.1 Debug Build and Test please(https://github.com/dotnet/corefx/issues/11820)

**Comment by ericeil** [comment]
Anyone have anything to add before I merge this?

**Pull request merged**